pkg/tools/list/dep_parser_test.go

```diff
@@ -37,53 +37,189 @@ func TestFindPkgInfo_failed(t *testing.T) {
+	for _, testdata := range testImportDepParser {
 		t.Run(testdata.name, func(t *testing.T) {
-			depParser, err := NewImportDepParser(testdata.root, DepOption{Files: testdata.files})
+			depParser, err := newImportDepParser(testdata.root, DepOption{Files: testdata.files})
 			assert.Nil(t, err, "NewDepParser failed")
-			deps := depParser.ListUpstreamFiles()
-			assert.ElementsMatch(t, testdata.depFiles, deps)
+			deps := depParser.upstreamFiles()
+			assert.ElementsMatch(t, testdata.upStreams, deps)
 		})
 	}
 }
 
-func TestDepParser_affected(t *testing.T) {
-	for _, testdata := range testDepParser {
+func TestImportDepParser_ListDownstreamFiles(t *testing.T) {
+	for _, testdata := range testImportDepParser {
 		t.Run(testdata.name, func(t *testing.T) {
-			depParser, err := NewImportDepParser(testdata.root, DepOption{Files: testdata.files, ChangedPaths: testdata.changed})
+			depParser, err := newImportDepParser(testdata.root, DepOption{Files: testdata.files, UpStreams: testdata.changed})
 			assert.Nil(t, err, "NewDepParser failed")
-			affected := depParser.ListDownStreamFiles()
-			assert.ElementsMatch(t, testdata.affected, affected)
+			affected := depParser.downStreamFiles()
+			assert.ElementsMatch(t, testdata.downStreams, affected)
 		})
 	}
 }
 
-var testDepParser = []struct {
-	name     string
-	root     string
-	files    []string
-	depFiles []string
-	changed  []string
-	affected []string
+func BenchmarkImportDepParser_walkDownStream(b *testing.B) {
+	testdata := testImportDepParser[0]
+	depParser, err := newImportDepParser(testdata.root, DepOption{Files: testdata.files, UpStreams: testdata.changed})
+	if err == nil {
+		for i := 0; i < b.N; i++ {
+			_ = depParser.downStreamFiles()
+		}
+	}
+}
+
+func BenchmarkImportDepParser_walkUpStreamFiles(b *testing.B) {
+	testdata := testImportDepParser[0]
+	depParser, err := newImportDepParser(testdata.root, DepOption{Files: testdata.files, UpStreams: testdata.changed})
+	if err == nil {
+		for i := 0; i < b.N; i++ {
+			_ = depParser.upstreamFiles()
+		}
+	}
+}
+
+func TestImportDepParser_fixImportPath(t *testing.T) {
+	testData := []struct {
+		name       string
+		filePath   string
+		importPath string
+		expect     string
+	}{
+		{
+			name:       "absolute import",
+			filePath:   "main.k",
+			importPath: "base.b",
+			expect:     "base/b",
+		},
+		{
+			name:       "relative import1",
+			filePath:   "base/b.k",
+			importPath: ".a",
+			expect:     "base/a",
+		},
+		{
+			name:       "relative import2",
+			filePath:   "base/a.k",
+			importPath: "..frontend",
+			expect:     "frontend",
+		},
+		{
+			name:       "invalid import: out of program bound",
+			filePath:   "base/a.k",
+			importPath: "...frontend",
+			expect:     "frontend",
+		},
+	}
+	for _, tData := range testData {
+		t.Run(tData.name, func(t *testing.T) {
+			assert.Equal(t, tData.expect, fixImportPath(tData.filePath, tData.importPath))
+		})
+	}
+}
+
+func TestImportDepParser_fixPath(t *testing.T) {
```

Comment:
s/`TestImportDepParser_fixPath`/`Test_fixPath`: fixPath is not the `ImportDepParser` method